### PR TITLE
[3.7] bpo-34247: Fix Python 3.7 initialization

### PR DESCRIPTION
**Pull request by vstinner**
* -X dev: it is now possible to override the memory allocator using
  PYTHONMALLOC even if the developer mode is enabled.
* Add _Py_InitializeFromConfig()
* Add _Py_Initialize_ReadEnvVars() to set global configuration
  variables from environment variables
* Fix the code to initialize Python: Py_Initialize() now also reads
  environment variables
* _Py_InitializeCore() can now be called repeatedly: the second
  and subsequent calls only replace the configuration, they don't
  recreate the main interpreter
* Write unit tests on Py_Initialize() and the different ways to
  configure Python
* The isolated mode now always sets Py_IgnoreEnvironmentFlag and
  Py_NoUserSiteDirectory to 1.
* pymain_read_conf() now saves/restores the configuration
  if the encoding changed

https://bugs.python.org/issue34247

### CODE REVIEW

**Comment by vstinner** [comment]
> bedevere/backport-pr — Not a valid backport PR title. 

Stupid bot. It doesn't like "[3.7] bpo-34247: Fix Python initialization" nor "bpo-34247: Fix Python 3.7 initialization". Hopefully, it doesn't prevent to merge the PR :-)

**Comment by vstinner** [comment]
@ronaldoussoren: I will merge this change in two weeks if I don't hear anything from you, since I know feel guilty of having introducing regressions in Python 3.7. This change adds new tests which is something very new for this part of the C API. The tests make me confident in my change :-) They helped me to fix bugs in this PR :-)

**Comment by vstinner** [comment]
This PR backports the best enhancements from the master branch without backporting everything, since I made too many changes in the master. And it seems that master is going to get much more changes, according to the PEP 432 discussion on python-dev. So I prefer to implement the *minimum* changes for 3.7 just to handle environment variables and to have working unit tests.

**Comment by ncoghlan** [comment]
Seeing if I can make the backport-pr bot happy by updating the PR title to refer to itself :)

File: Include/pylifecycle.h
Comment (ncoghlan) [inline Include/pylifecycle.h]: We're already inside a Py_LIMITED_API guard here, so this isn't needed.

File: Misc/NEWS.d/next/Core and Builtins/2018-08-03-21-59-06.bpo-34170.v1h_H2.rst
Comment (ncoghlan) [inline Misc/NEWS.d/next/Core and Builtins/2018-08-03-21-59-06.bpo-34170.v1h_H2.rst:2]: We also need a NEWS entry for the bpo-34247 fix

File: Modules/main.c
Comment (ncoghlan) [inline Modules/main.c]: Something like `_Py_Initialize_ReadEnvVarsNoAlloc` might be a better name here, as this doesn't read *all* the env vars, only the ones that it can process without needing to allocate any memory to store the results.

With the current name, it gives the impression it reads all the environment variables, when it actually skips over the more complex ones like `PYTHONPATH`.

File: Python/pylifecycle.c
Comment (ncoghlan) [inline Python/pylifecycle.c]: This error message isn't right for the reconfiguration case. Perhaps something like "Core reconfiguration: no main interpreter found".

Then the thread state lookup above could also gain the "Core reconfiguration: " prefix.

**Review by ncoghlan — COMMENTED** [review]
This is looking pretty good to me - just a few readability/debuggability comments inline.

**Comment by ncoghlan** [comment]
After reviewing the change, I slightly reworded the section in the commit message about `_Py_InitializeCore`

File: Lib/test/test_embed.py
Comment (matrixise) [inline Lib/test/test_embed.py:297]: I am curious, why do you need to convert the dict to a list? usually, you will iterate over the keys, is there a subtlety?

Comment (vstinner) [reply Lib/test/test_embed.py:297]: Because the loop body modifies the env dict.

**Comment by Mariatta** [comment]
bedevere/backport-pr status is not required and does not block merging. You'll notice the "Squash and merge" button is still enabled.

**Comment by Mariatta** [comment]
And the PR needs to start with `[3.7]`.

**Comment by ncoghlan** [comment]
Ah, cool - that means the self-reference *is* a valid workaround, I just missed that the prefix was missing as well.

**Comment by Mariatta** [comment]
I've deployed the latest change in bedevere, so now it is only looking for the `[X.Y]` prefix, and it does not care about the `(GH-NNNN)` anymore.
But this status is now "Required".

Example:
screen shot 2018-08-04 at 8 08 33 am

File: Include/pylifecycle.h
Comment (vstinner) [reply Include/pylifecycle.h]: fixed

File: Misc/NEWS.d/next/Core and Builtins/2018-08-03-21-59-06.bpo-34170.v1h_H2.rst
Comment (vstinner) [reply Misc/NEWS.d/next/Core and Builtins/2018-08-03-21-59-06.bpo-34170.v1h_H2.rst:2]: done

File: Modules/main.c
Comment (vstinner) [reply Modules/main.c]: function renamed

File: Python/pylifecycle.c
Comment (vstinner) [reply Python/pylifecycle.c]: I changed the error message to "no main interpreter found". The displayed error message contains the name of the current function _Py_InitializeCore_impl() thanks to _Py_INIT_ERR() magic. I don't want to add a prefix per error message. Such error should not occur in practice anyway.

**Review by vstinner — COMMENTED** [review]
I addressed all @ncoghlan's comment in my new commit 6c3b5ec28b5cd5cb72cded83ed530f2dbac5952b.

**Review by ncoghlan — APPROVED** [review]
+1 for going ahead with this version.

We may want to tweak the porting note in What's New, as I believe this may read more environment variables at init time than the interpreter did previously, so embedding applications that want more complete control may now need to set Py_IgnoreEnvironmentFlag when they could previously get by without it.

**Pull request merged**
